# KUIS 2 Pembelajaran Mesin

## Tentang dataset

Pada sebuah lembaga kemanusiaan berbasis teknologi informasi, dibentuk beberapa tim kemanusiaan yang bertugas untuk terjun ke masyarakat memberikan bantuan dalam bentuk aplikasi komputer atau layanan pelatihan IT.  Kegiatn tersebut dilaksanakan setiap tahun dengan ketentuan anggota organisasi tersebut dapat bergabung ke beberapa tim sebagai ketua atau anggota. Terdapat pula anggota yang sudah tidak aktif di organisasi tersebut (terhapus di database mereka —> label N/A).  Organisasi tersebut juga melakukan rekrutmen volunteer, namun tidak semua tim menggunakan bantuan volunteer. Hasil kegiatan dari setiap tim dapat diwujudkan ke dalam sebuah pengakuan hak cipta atau publikasi karya ilmiah atau publikasi berita di media massa.

## Ketentuan
1. Pengerjaan secara individu alias setiap mahasiswa!
2. Terdapat data dengan nilai #N/A, pertimbangkan cara terbaik untuk mengatasi nilai tersebut.
3. Tidak boleh ada duplikasi pekerjaan antar mahasiswa
4. Berikan penjelasan untuk setiap tahapan. Jangan hanya kode!

## Soal
1. Lakukan pemilahan data training dan testing yang sesuai dan proporsional dari data tersebut! (10 poin)
2. Tentukan data validasi! (10 poin)
3. Tentukan secara tepat kolom yang menjadi "label" (20 poin)
4. Terapkan dua jenis metode klasifikasi dan lakukan perbandingan akurasi  dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran yang dihasilkan! (20 poin)
5. Lakukan tunning hyperparameter berdasarkan hasil klasifikasi pada nomor 3 menggunakan metode Grid Search dan Random Search! Simpulkan parameter apa yang paling baik digunakan berdasarkan metode klasifikasi yang dipilih, dan berapa akurasinya. (40 poin)

# Identitas

NIM: 2041720108

Nama: Raihan Hidayatullah Djunaedi

Kelas: TI-3A

## Jawab

### Import

In [85]:
# Import untuk Eliminating Warning
import warnings
warnings.filterwarnings("ignore")
from sklearn.exceptions import FitFailedWarning 

import pandas as pd # Import pandas
import numpy as np # Import numpy
from scipy.stats import mode # Import mode untuk mengisi data nan dengan modus
from sklearn.model_selection import train_test_split # Import train test split
from sklearn.preprocessing import LabelEncoder # Import Label Encoder
from sklearn.naive_bayes import MultinomialNB # import MultinomialNB
from sklearn.ensemble import RandomForestClassifier # import RandomForest
from sklearn.metrics import accuracy_score # import accuracy score
from sklearn.tree import DecisionTreeClassifier # import DT

### Ketentuan

In [86]:
# Read Data
df = pd.read_excel('data/dataset_quiz.xlsx')

# Menampilkan data teratas jumlah 10
df.head(10)

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0    7206.0       7214.0       6211.0       8602.0       7801.0          NaN   
1    7214.0       7109.0       8402.0       7903.0       9202.0          NaN   
2    5911.0       8201.0       8701.0       9101.0       9003.0          NaN   
3    8402.0       7702.0       8602.0       8903.0       8903.0          NaN   
4    8201.0       8305.0       5911.0       5903.0       8402.0          NaN   
5    8008.0       7607.0       7903.0       7702.0       8905.0          NaN   
6    8402.0       8803.0       9102.0       8804.0       8005.0          NaN   
7    8101.0       7702.0       7308.0       8602.0       9003.0          NaN   
8       NaN       8805.0       8807.0       8803.0       9003.0          NaN   
9    8805.0       7206.0       7903.0       7702.0       9002.0          NaN   

  keterlibatan_vltr  jml_vltr_terlibat  tahun    skema luaran_hki  \
0             TIDAK                  0   2019  REGULER  TIDAK ADA   
1             TIDAK                  0   2019  REGULER  TIDAK ADA   
2             TIDAK                  0   2019  REGULER  TIDAK ADA   
3             TIDAK                  0   2019  REGULER  TIDAK ADA   
4             TIDAK                  0   2019  REGULER  TIDAK ADA   
5             TIDAK                  0   2019  REGULER  TIDAK ADA   
6             TIDAK                  0   2019  REGULER  TIDAK ADA   
7             TIDAK                  0   2019  REGULER  TIDAK ADA   
8             TIDAK                  0   2019  REGULER  TIDAK ADA   
9             TIDAK                  0   2019  REGULER  TIDAK ADA   

  jenis_luaran_hki potensi_hki luaran_lain jenis_luaran_lain  \
0        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
1        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
2        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
3        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
4        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
5        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
6        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
7        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
8        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
9        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   

  status_luaran_lain  
0          TIDAK ADA  
1          TIDAK ADA  
2          TIDAK ADA  
3          TIDAK ADA  
4          TIDAK ADA  
5          TIDAK ADA  
6          TIDAK ADA  
7          TIDAK ADA  
8          TIDAK ADA  
9          TIDAK ADA

In [87]:
# Cek jumlah kolom null
df.isnull().sum()

id_ketua               3
id_anggota1            5
id_anggota2            4
id_anggota3            3
id_anggota4           13
id_anggota5           84
keterlibatan_vltr      0
jml_vltr_terlibat      0
tahun                  0
skema                  0
luaran_hki             0
jenis_luaran_hki       0
potensi_hki            0
luaran_lain            0
jenis_luaran_lain      0
status_luaran_lain     0
dtype: int64

In [88]:
# # Mengisi kolom null menjadi 0

# df['id_ketua'] = df['id_ketua'].replace(np.nan, 0)
# df['id_anggota1'] = df['id_anggota1'].replace(np.nan, 0)
# df['id_anggota2'] = df['id_anggota2'].replace(np.nan, 0)
# df['id_anggota3'] = df['id_anggota3'].replace(np.nan, 0)
# df['id_anggota4'] = df['id_anggota4'].replace(np.nan, 0)
# df['id_anggota5'] = df['id_anggota5'].replace(np.nan, 0)

# Mengisi kolom null dengan modus

df['id_ketua'] = df['id_ketua'].fillna(df['id_ketua'].mode()[0])
df['id_anggota1'] = df['id_anggota1'].fillna(df['id_anggota1'].mode()[0])
df['id_anggota2'] = df['id_anggota2'].fillna(df['id_anggota2'].mode()[0])
df['id_anggota3'] = df['id_anggota3'].fillna(df['id_anggota3'].mode()[0])
df['id_anggota4'] = df['id_anggota4'].fillna(df['id_anggota4'].mode()[0])
df['id_anggota5'] = df['id_anggota5'].fillna(df['id_anggota5'].mode()[0])

In [89]:
# Cek kolom null kembali
df.isnull().sum()

id_ketua              0
id_anggota1           0
id_anggota2           0
id_anggota3           0
id_anggota4           0
id_anggota5           0
keterlibatan_vltr     0
jml_vltr_terlibat     0
tahun                 0
skema                 0
luaran_hki            0
jenis_luaran_hki      0
potensi_hki           0
luaran_lain           0
jenis_luaran_lain     0
status_luaran_lain    0
dtype: int64

In [90]:
# Mengencode label yang belum terencode agar dapat di train dan dicari akurasinya
# Inisiasi label encoder
e = LabelEncoder()

# Menerapkan label encoder
df['keterlibatan_vltr'] = e.fit_transform(df['keterlibatan_vltr'])
df['tahun'] = e.fit_transform(df['tahun'])
df['skema'] = e.fit_transform(df['skema'])
df['luaran_hki'] = e.fit_transform(df['luaran_hki'])
df['jenis_luaran_hki'] = e.fit_transform(df['jenis_luaran_hki'])
df['potensi_hki'] = e.fit_transform(df['potensi_hki'])
df['luaran_lain'] = e.fit_transform(df['luaran_lain'])
df['jenis_luaran_lain'] = e.fit_transform(df['jenis_luaran_lain'])
df['status_luaran_lain'] = e.fit_transform(df['status_luaran_lain'])

# Cek hasil
df.head(10)

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0    7206.0       7214.0       6211.0       8602.0       7801.0       8812.0   
1    7214.0       7109.0       8402.0       7903.0       9202.0       8812.0   
2    5911.0       8201.0       8701.0       9101.0       9003.0       8812.0   
3    8402.0       7702.0       8602.0       8903.0       8903.0       8812.0   
4    8201.0       8305.0       5911.0       5903.0       8402.0       8812.0   
5    8008.0       7607.0       7903.0       7702.0       8905.0       8812.0   
6    8402.0       8803.0       9102.0       8804.0       8005.0       8812.0   
7    8101.0       7702.0       7308.0       8602.0       9003.0       8812.0   
8    8402.0       8805.0       8807.0       8803.0       9003.0       8812.0   
9    8805.0       7206.0       7903.0       7702.0       9002.0       8812.0   

   keterlibatan_vltr  jml_vltr_terlibat  tahun  skema  luaran_hki  \
0                  0                  0      0      1           1   
1                  0                  0      0      1           1   
2                  0                  0      0      1           1   
3                  0                  0      0      1           1   
4                  0                  0      0      1           1   
5                  0                  0      0      1           1   
6                  0                  0      0      1           1   
7                  0                  0      0      1           1   
8                  0                  0      0      1           1   
9                  0                  0      0      1           1   

   jenis_luaran_hki  potensi_hki  luaran_lain  jenis_luaran_lain  \
0                 1            2            0                  3   
1                 1            2            0                  3   
2                 1            2            0                  3   
3                 1            2            0                  3   
4                 1            2            0                  3   
5                 1            2            0                  3   
6                 1            2            0                  3   
7                 1            2            0                  3   
8                 1            2            0                  3   
9                 1            2            0                  3   

   status_luaran_lain  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1  
5                   1  
6                   1  
7                   1  
8                   1  
9                   1

In [91]:
# Menampilkan Data yang sudah tidak null 
df.head(10)

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0    7206.0       7214.0       6211.0       8602.0       7801.0       8812.0   
1    7214.0       7109.0       8402.0       7903.0       9202.0       8812.0   
2    5911.0       8201.0       8701.0       9101.0       9003.0       8812.0   
3    8402.0       7702.0       8602.0       8903.0       8903.0       8812.0   
4    8201.0       8305.0       5911.0       5903.0       8402.0       8812.0   
5    8008.0       7607.0       7903.0       7702.0       8905.0       8812.0   
6    8402.0       8803.0       9102.0       8804.0       8005.0       8812.0   
7    8101.0       7702.0       7308.0       8602.0       9003.0       8812.0   
8    8402.0       8805.0       8807.0       8803.0       9003.0       8812.0   
9    8805.0       7206.0       7903.0       7702.0       9002.0       8812.0   

   keterlibatan_vltr  jml_vltr_terlibat  tahun  skema  luaran_hki  \
0                  0                  0      0      1           1   
1                  0                  0      0      1           1   
2                  0                  0      0      1           1   
3                  0                  0      0      1           1   
4                  0                  0      0      1           1   
5                  0                  0      0      1           1   
6                  0                  0      0      1           1   
7                  0                  0      0      1           1   
8                  0                  0      0      1           1   
9                  0                  0      0      1           1   

   jenis_luaran_hki  potensi_hki  luaran_lain  jenis_luaran_lain  \
0                 1            2            0                  3   
1                 1            2            0                  3   
2                 1            2            0                  3   
3                 1            2            0                  3   
4                 1            2            0                  3   
5                 1            2            0                  3   
6                 1            2            0                  3   
7                 1            2            0                  3   
8                 1            2            0                  3   
9                 1            2            0                  3   

   status_luaran_lain  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1  
5                   1  
6                   1  
7                   1  
8                   1  
9                   1

In [92]:
# Cek df apakah tetap sama

display(df.head())
print(f'Dimensi data Quiz: {df.shape}')

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0    7206.0       7214.0       6211.0       8602.0       7801.0       8812.0   
1    7214.0       7109.0       8402.0       7903.0       9202.0       8812.0   
2    5911.0       8201.0       8701.0       9101.0       9003.0       8812.0   
3    8402.0       7702.0       8602.0       8903.0       8903.0       8812.0   
4    8201.0       8305.0       5911.0       5903.0       8402.0       8812.0   

   keterlibatan_vltr  jml_vltr_terlibat  tahun  skema  luaran_hki  \
0                  0                  0      0      1           1   
1                  0                  0      0      1           1   
2                  0                  0      0      1           1   
3                  0                  0      0      1           1   
4                  0                  0      0      1           1   

   jenis_luaran_hki  potensi_hki  luaran_lain  jenis_luaran_lain  \
0                 1            2            0                  3   
1                 1            2            0                  3   
2                 1            2            0                  3   
3                 1            2            0                  3   
4                 1            2            0                  3   

   status_luaran_lain  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1

Dimensi data Quiz: (85, 16)


In [93]:
# Menampilkan deskripsi data
df.describe()

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  \
count    85.000000    85.000000    85.000000    85.000000    85.000000   
mean   7970.670588  8123.858824  8248.929412  8425.505882  8622.647059   
std     854.472664  1150.092182  1439.784415  1083.076445  1052.686729   
min    5903.000000   569.000000   569.000000   569.000000   569.000000   
25%    7706.000000  7607.000000  7903.000000  8101.000000  8602.000000   
50%    8201.000000  8305.000000  8704.000000  8702.000000  8901.000000   
75%    8602.000000  8901.000000  8907.000000  8905.000000  9003.000000   
max    9209.000000  9402.000000  9402.000000  9303.000000  9303.000000   

       id_anggota5  keterlibatan_vltr  jml_vltr_terlibat      tahun  \
count         85.0          85.000000          85.000000  85.000000   
mean        8812.0           0.470588           1.341176   1.611765   
std            0.0           0.502096           1.452034   1.155670   
min         8812.0           0.000000           0.000000   0.000000   
25%         8812.0           0.000000           0.000000   1.000000   
50%         8812.0           0.000000           0.000000   2.000000   
75%         8812.0           1.000000           3.000000   3.000000   
max         8812.0           1.000000           4.000000   3.000000   

           skema  luaran_hki  jenis_luaran_hki  potensi_hki  luaran_lain  \
count  85.000000   85.000000         85.000000     85.00000    85.000000   
mean    0.941176    0.905882          0.905882      2.00000     0.188235   
std     0.236691    0.293725          0.293725      0.46291     0.393220   
min     0.000000    0.000000          0.000000      0.00000     0.000000   
25%     1.000000    1.000000          1.000000      2.00000     0.000000   
50%     1.000000    1.000000          1.000000      2.00000     0.000000   
75%     1.000000    1.000000          1.000000      2.00000     0.000000   
max     1.000000    1.000000          1.000000      3.00000     1.000000   

       jenis_luaran_lain  status_luaran_lain  
count          85.000000           85.000000  
mean            2.564706            1.129412  
std             0.993396            0.402026  
min             0.000000            0.000000  
25%             3.000000            1.000000  
50%             3.000000            1.000000  
75%             3.000000            1.000000  
max             3.000000            2.000000

### Lakukan pemilahan data training dan testing yang sesuai dan proporsional dari data tersebut! (10 poin)

In [94]:
# Split data
# Split data training dan dan lainnya
# Rasio yang akan kita gunakan adalah 8:1:1
df_train, df_test = train_test_split(df, test_size=0.2, random_state=0)

# Cek masing-masing ukuran data
print(f'Jumlah data asli: {df.shape[0]}')
print(f'Jumlah data train: {df_train.shape[0]}')
print(f'Jumlah data test: {df_test.shape[0]}')

# Cek rasio tiap label
print('=========')
print(f'Jumlah label data asli:\n{df.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data train:\n{df_train.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data test:\n{df_test.keterlibatan_vltr.value_counts()}')

Jumlah data asli: 85
Jumlah data train: 68
Jumlah data test: 17
Jumlah label data asli:
0    45
1    40
Name: keterlibatan_vltr, dtype: int64
Jumlah label data train:
0    36
1    32
Name: keterlibatan_vltr, dtype: int64
Jumlah label data test:
0    9
1    8
Name: keterlibatan_vltr, dtype: int64


### Menentukan data validasi! (10 poin)

In [95]:
# Split data
# Split data training dan dan lainnya
# data lainnya, akan kita split lagi menjadi validasi dan testing.
# Rasio yang akan kita gunakan adalah 8:1:1
df_train, df_unseen = train_test_split(df, test_size=0.2, random_state=0, stratify=df['keterlibatan_vltr'])

# Split lagi antara validasi dan testing
df_val, df_test = train_test_split(df_unseen, test_size=0.5, random_state=0, stratify=df_unseen['keterlibatan_vltr'])

# Cek masing-masing ukuran data

print(f'Jumlah data asli: {df.shape[0]}')
print(f'Jumlah data train: {df_train.shape[0]}')
print(f'Jumlah data val: {df_val.shape[0]}')
print(f'Jumlah data test: {df_test.shape[0]}')

# Cek rasio tiap label
print('=========')
print(f'Jumlah label data asli:\n{df.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data train:\n{df_train.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data val:\n{df_val.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data test:\n{df_test.keterlibatan_vltr.value_counts()}')

Jumlah data asli: 85
Jumlah data train: 68
Jumlah data val: 8
Jumlah data test: 9
Jumlah label data asli:
0    45
1    40
Name: keterlibatan_vltr, dtype: int64
Jumlah label data train:
0    36
1    32
Name: keterlibatan_vltr, dtype: int64
Jumlah label data val:
0    4
1    4
Name: keterlibatan_vltr, dtype: int64
Jumlah label data test:
0    5
1    4
Name: keterlibatan_vltr, dtype: int64


### Menentukan secara tepat kolom yang menjadi "label" (20 poin)

In [96]:
# Seleksi Fitur
X = df.iloc[:, 6:]
y = df['keterlibatan_vltr']

# Cek jumlah fitur dan instance
X.shape

(85, 10)

In [97]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

### Terapkan dua jenis metode klasifikasi dan lakukan perbandingan akurasi  dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran yang dihasilkan! (20 poin)

DecisionTree

In [98]:
# Mendefinisikan model Decision Tree Classifier
dt = DecisionTreeClassifier()

# Sesuaikan dt ke set training
dt.fit(X_train, y_train)

# Memprediksi label set test
y_pred_dt = dt.predict(X_test)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Test set accuracy: 1.00
Test set accuracy: 1.0


Random Forest

In [99]:
# Mendefinisikan model Random Forest Classifier
rf = RandomForestClassifier()

# Sesuaikan rf ke set training
rf.fit(X_train, y_train)

# Memprediksi label set test
y_pred_rf = rf.predict(X_test)

#  menghitung set accuracy
acc_rf = accuracy_score(y_test, y_pred_rf)
print("Test set accuracy: {:.2f}".format(acc_rf))
print(f"Test set accuracy: {acc_rf}")

Test set accuracy: 1.00
Test set accuracy: 1.0


Naive Bayess

In [100]:
# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Fit model
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_pred_train = mnb.predict(X_train)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_pred_train)

# Prediksi dengan data training
y_pred_test = mnb.predict(X_test)

# Evaluasi akurasi data training
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Hasil akurasi data train: 0.9852941176470589
Hasil akurasi data test: 1.0


### Lakukan tunning hyperparameter berdasarkan hasil klasifikasi pada nomor 3 menggunakan metode Grid Search dan Random Search! Simpulkan parameter apa yang paling baik digunakan berdasarkan metode klasifikasi yang dipilih, dan berapa akurasinya.

In [101]:
from sklearn.model_selection import RepeatedKFold

# inisiasi repated k-fold
cv = RepeatedKFold(n_splits=4, n_repeats=3, random_state=42)

#### Grid Search

Decission Tree

In [102]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# inisiasi model
dt = DecisionTreeClassifier()

# Definisikan hyperparameter yang akan digunakan
# sklearn menerima dalam bentuk dictionary
# nama hyperparamater HARUS SESUAI dengan dokumentasi sklearn
params = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': list(range(5,11)),
    
}

# inisiasi grid berdasarkan nilai repeated k-fold dan hyperparameter
grid = GridSearchCV(dt, param_grid=params, cv=cv)

# Fit / latih berdasarkan grid
# %timeit merupakan magic command didalam ipython notebook
# yang dapat kita gunakan untuk menghitung waktu komputasi
# cara ini cukup efektif untuk melakukan evaluasi suatu algoritma atau prosedur
%timeit grid.fit(X, y)

# Evaluasi dengan score
score = grid.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {grid.best_params_}')
print(f'Best Estimator: {grid.best_estimator_}')
print(f'Best Score: {grid.best_score_}')

879 ms ± 40.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'criterion': 'gini', 'max_depth': 5}
Best Estimator: DecisionTreeClassifier(max_depth=5)
Best Score: 0.996031746031746


Random Forest

In [103]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# inisiasi model
rf = RandomForestClassifier()

# Definisikan hyperparameter yang akan digunakan
# sklearn menerima dalam bentuk dictionary
# nama hyperparamater HARUS SESUAI dengan dokumentasi sklearn
params = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': list(range(5,11)),
}

# inisiasi grid berdasarkan nilai repeated k-fold dan hyperparameter
grid = GridSearchCV(rf, param_grid=params, cv=cv)

# Fit / latih berdasarkan grid
# %timeit merupakan magic command didalam ipython notebook
# yang dapat kita gunakan untuk menghitung waktu komputasi
# cara ini cukup efektif untuk melakukan evaluasi suatu algoritma atau prosedur
%timeit grid.fit(X, y)

# Evaluasi dengan score
score = grid.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {grid.best_params_}')
print(f'Best Estimator: {grid.best_estimator_}')
print(f'Best Score: {grid.best_score_}')

17.9 s ± 1.63 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'criterion': 'gini', 'max_depth': 7}
Best Estimator: RandomForestClassifier(max_depth=7)
Best Score: 1.0


#### Random Search

Decission Tree

In [104]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV

# inisiasi model
dt = DecisionTreeClassifier()

# Definisikan hyperparameter yang akan digunakan
# sklearn menerima dalam bentuk dictionary
# nama hyperparamater HARUS SESUAI dengan dokumentasi sklearn
rs_space = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': list(range(5,11))
}

# inisiasi random berdasarkan nilai repeated k-fold dan hyperparameter
random = RandomizedSearchCV(dt, rs_space, cv=cv, scoring='accuracy')

# Fit / latih berdasarkan random
# %timeit merupakan magic command didalam ipython notebook
# yang dapat kita gunakan untuk menghitung waktu komputasi
# cara ini cukup efektif untuk melakukan evaluasi suatu algoritma atau prosedur
%timeit random.fit(X, y)

# Evaluasi dengan score
score = random.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {random.best_params_}')
print(f'Best Estimator: {random.best_estimator_}')
print(f'Best Score: {random.best_score_}')

407 ms ± 28.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'max_depth': 9, 'criterion': 'gini'}
Best Estimator: DecisionTreeClassifier(max_depth=9)
Best Score: 0.996031746031746


Random Forest

In [105]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV

# inisiasi model
rf = RandomForestClassifier()

# Definisikan hyperparameter yang akan digunakan
# sklearn menerima dalam bentuk dictionary
# nama hyperparamater HARUS SESUAI dengan dokumentasi sklearn
rs_space = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': list(range(5,11))
}

# inisiasi random berdasarkan nilai repeated k-fold dan hyperparameter
random = RandomizedSearchCV(rf, rs_space, cv=cv, scoring='accuracy')

# Fit / latih berdasarkan random
# %timeit merupakan magic command didalam ipython notebook
# yang dapat kita gunakan untuk menghitung waktu komputasi
# cara ini cukup efektif untuk melakukan evaluasi suatu algoritma atau prosedur
%timeit random.fit(X, y)

# Evaluasi dengan score
score = random.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {random.best_params_}')
print(f'Best Estimator: {random.best_estimator_}')
print(f'Best Score: {random.best_score_}')

10.4 s ± 1.13 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'max_depth': 10, 'criterion': 'gini'}
Best Estimator: RandomForestClassifier(max_depth=10)
Best Score: 0.996031746031746


## Kesimpulan



Grid Search : 
- Decission Tree
  - Hasil evaluasi: 1.0
  - Konfigurasi hyperparameter: {'criterion': 'gini', 'max_depth': 5}
  - Best Estimator: DecisionTreeClassifier(max_depth=5)
  - Best Score: 0.996031746031746

- Random Forest
  - Hasil evaluasi: 1.0
  - Konfigurasi hyperparameter: {'criterion': 'gini', 'max_depth': 7}
  - Best Estimator: RandomForestClassifier(max_depth=7)
  - Best Score: 1.0

Random Search : 
- Decission Tree
  - Hasil evaluasi: 1.0
  - Konfigurasi hyperparameter: {'max_depth': 9, 'criterion': 'gini'}
  - Best Estimator: DecisionTreeClassifier(max_depth=9)
  - Best Score: 0.996031746031746

- Random Forest
  - Hasil evaluasi: 1.0
  - Konfigurasi hyperparameter: {'max_depth': 10, 'criterion': 'gini'}
  - Best Estimator: RandomForestClassifier(max_depth=10)
  - Best Score: 0.996031746031746

Random Search lebih stabil daripada grid search